### PR TITLE
Implement home-manager support

### LINKED ISSUE #62: Does `nix-sops` support using it as a module in `home-manager` configuration?
If it is currently supported, please offer documentation about how to set it up.

If it is currently *not* supported, please consider supporting it.

### LINKED ISSUE #163: nix-darwin support
Tangentially related to #62, it would be useful to be able to ship secrets on nix-darwin systems.

### PR DESCRIPTION
**Pull request by dasJ**
Closes #62
Closes #163

### CODE REVIEW

File: README.md
Comment (Mic92) [inline README.md:757]: `$XDG_RUNTIME_DIR` can be not set on non-nixos. We should recommend to have it point it to a tmpfs so that secrets are not persisted stored in plain text or otherwise live with the consequences...

Comment (Mic92) [inline README.md:757]: `$XDG_RUNTIME_DIR` can be not set on non-nixos. We should recommend to have it point it to a tmpfs so that secrets are not persisted stored in plain text or otherwise live with the consequences...

File: README.md
Comment (Mic92) [inline README.md:764]: Maybe fetchTarball could work here? I would also expose the module via flake.nix.

Comment (Mic92) [inline README.md:764]: Maybe fetchTarball could work here? I would also expose the module via flake.nix.

File: README.md
Comment (Mic92) [inline README.md:770]: It's confusing that home-manager and nixos configuration are in one code block. I would also keep the `home-manager` configuration below separate from importing sops-nix.

Comment (Mic92) [inline README.md:770]: It's confusing that home-manager and nixos configuration are in one code block. I would also keep the `home-manager` configuration below separate from importing sops-nix.

File: README.md
Comment (Mic92) [inline README.md:777]: In that case we should also replace `%%` with `%` without expanding it for escaping purpose.

```suggestion
      path = "%r/test.txt"; # %r gets replaced with your $XDG_RUNTIME_DIR
```

Comment (Mic92) [inline README.md:777]: In that case we should also replace `%%` with `%` without expanding it for escaping purpose.

```suggestion
      path = "%r/test.txt"; # %r gets replaced with your $XDG_RUNTIME_DIR
```

File: README.md
Comment (Mic92) [inline README.md:774]: I am inclined to make normal age keys the default here as a best practice. Most of the time you want to have ssh keys encrypted because of their actual use.

Comment (Mic92) [inline README.md:774]: I am inclined to make normal age keys the default here as a best practice. Most of the time you want to have ssh keys encrypted because of their actual use.

File: pkgs/sops-install-secrets/main.go
Comment (Mic92) [inline pkgs/sops-install-secrets/main.go:890]: if `XDG_RUNTIME_DIR` is not set, it's likely that there is no `/run/user/<uid>`. I would rather return an error here rather than writing to some random directory.

Comment (Mic92) [inline pkgs/sops-install-secrets/main.go:890]: if `XDG_RUNTIME_DIR` is not set, it's likely that there is no `/run/user/<uid>`. I would rather return an error here rather than writing to some random directory.

File: modules/home-manager/sops.nix
Comment (Mic92) [inline modules/home-manager/sops.nix:205]: We should have documentation on how to have services depending on secrets waiting for this service.

Comment (Mic92) [inline modules/home-manager/sops.nix:205]: We should have documentation on how to have services depending on secrets waiting for this service.

**Comment by Kranzes** [comment]
The decryption of the keys should be part of home `home.activation`.

**Comment by Mic92** [comment]
> The decryption of the keys should be part of `home.activation`.

`home.activation` is not run on boot, which means we also need a service to decrypt secrets to tmpfs.
Because of this we already need to have dependencies between services, I don't see the value added by also having it in activation phase. Does this make sense?

**Comment by Kranzes** [comment]
I did some further testing yesterday and it did start on ("boot") login to the user.

**Comment by Mic92** [comment]
> I did some further testing yesterday and it did start on ("boot") login to the user.

Ok. Than I need to investigate this.

**Comment by Kranzes** [comment]
I recommend you to test it as well. also I did some work on agenix home-manager module yesterday, https://github.com/ryantm/agenix/pull/109, you might want to take a look. Also I mentioned all the challenges I encountered on the agenix matrix channel if you want to go and read them.

**Comment by NobbZ** [comment]
@Kranzes 

> I did some further testing yesterday and it did start on ("boot") login to the user.

AFAIU this is only true when using `home-manager` as a module for nixos, not when used standalone. As activation is implemented as a user service on nixos then.

Therefore I think using services which are `Before` the user activation if and only if used as a service, but run standalone on standalone.

I have no clue how to solve this on darwin though. I do not have any experience with that system, I just know that there is no equivalent to systemd user scervices.

**Comment by mtoohey31** [comment]
> I have no clue how to solve this on darwin though. I do not have any experience with that system, I just know that there is no equivalent to systemd user scervices.

Darwin has launchd, and it looks like home-manager has [a module](https://github.com/nix-community/home-manager/tree/master/modules/launchd) for it, so that might be a good solution.

**Comment by otavio** [comment]
Looking forward to use this in my nix-config :-D

**Comment by dasJ** [comment]
@mtoohey31 could you help with the launchd logic? I don't really have the infrastructure to test anything in that regard

**Comment by dasJ** [comment]
I think I have pretty much all review comments fixed now.
Still missing is launchd support

**Comment by dasJ** [comment]
Added an untested launchd module, please test

**Comment by mtoohey31** [comment]
> Added an untested launchd module, please test

Sorry for the slow response, I'd be happy to help with testing but I'm out of time tonight. I'll try to get to it soon.

File: README.md
Comment (Mic92) [inline README.md:737]: Reminder to myself to expose this in the sops flake as well.

File: modules/home-manager/sops.nix
Comment (Mic92) [inline modules/home-manager/sops.nix:180]:
```suggestion
        example = "/home/someuser/.gnupg";
```

File: modules/home-manager/sops.nix
Comment (Mic92) [inline modules/home-manager/sops.nix:150]: I would point in the example to somewhere in `/home` maybe.

File: README.md
Comment (Mic92) [inline README.md:744]:
```suggestion
**Since the secrets are decrypted there, it's highly recommended to use a tmpfs for `$XDG_RUNTIME_DIR` to avoid storing secrets in plain text on persistent storage. Linux distributions using systemd-logind do that out-of-the-box.**
```

File: README.md
Comment (pogobanane) [inline README.md:771]:
```suggestion
    age.keyFile = "/home/user/.age-key.txt"; # must have no password!
```

File: README.md
Comment (pogobanane) [inline README.md:775]:
```suggestion
    defaultSopsFile = ./secrets.yml.enc;
    sops.secrets.test = {
      # sopsFile = ./secrets.yml.enc; # optionally define per-secret files
```

**Review by pogobanane — COMMENTED** [review]
I have to `systemctl --user restart sops-nix` manually after every change. Can this be automated?

File: pkgs/sops-install-secrets/main.go
Comment (pogobanane) [inline pkgs/sops-install-secrets/main.go:903]: I'm building my home-manager config with root as part of my nixos-config. `Nixos-rebuild switch` fails for me with:
```
error: builder for '/nix/store/bhscqh6gvv32cda919mkh2sa7yaa9chi-manifest.json.drv' failed with exit code 1;
       last 1 log lines:
       > /nix/store/ppl4vbv67c420zk9r87izdsfwms5q2ja-sops-install-secrets-0.0.1/bin/sops-install-secrets: $XDG_RUNTIME_DIR is not set!
       For full logs, run 'nix log /nix/store/bhscqh6gvv32cda919mkh2sa7yaa9chi-manifest.json.drv'.
error: 1 dependencies of derivation '/nix/store/a93kndskwslvnaaymqqk4svjhj7s33gg-sops-nix-user.drv' failed to build
```
I suspect this happens during the check phase. It may be sufficient to just ignore `$XDG_RUNTIME_DIR` during `--check-mode`.

Comment (pogobanane) [inline pkgs/sops-install-secrets/main.go:903]: this fixes my home-manager issue:
```suggestion
		if opts.checkMode == Off && !ok {
		        // Sometimes builds run check-mode outside the users environment. Thus
		        // this env var may not be set at check time, but it must be at runtime.
			return fmt.Errorf("$XDG_RUNTIME_DIR is not set!")
```

File: README.md
Comment (pogobanane) [inline README.md:774]:
```suggestion
    #age.sshKeyPaths = [ "/home/user/path-to-ssh-key" ];
    defaultSopsFile = ./secrets.yaml;
    sops.secrets.test = {
```

Comment (pogobanane) [reply README.md:774]: this is required, but got lost somehow

**Review by pogobanane — APPROVED** [review]
maybe wait for @mtoohey31 s feedback for macos, but you have my approval for nixos.

**Review by mtoohey31 — CHANGES_REQUESTED** [review]
Sorry that it's taken me a bit to get to this. In addition to the other review comment, `sops-install-secrets` requires some work before I can fully test this.

`default.nix` needs to be updated so `sops-install-secrets` is available for darwin, and `pkgs/sops-install-secrets/default.nix` needs to have `platforms.darwin` added to the supported platforms.

Getting `sops-install-secrets` to actually build on MacOS will require some work too. We need to remove the `// +build linux` constraints from `main.go` and `main_test.go`, and after that there are still some issues with uses of the `unix` package:

```
# github.com/Mic92/sops-nix/pkgs/sops-install-secrets
./main.go:119:54: undefined: unix.O_PATH
./main.go:133:44: undefined: unix.AT_EMPTY_PATH
./main.go:328:23: too many arguments in call to unix.Mount
./main.go:328:53: undefined: unix.MS_NODEV
./main.go:328:67: undefined: unix.MS_NOSUID
```

I believe we'll need Linux and MacOS-specific versions of the `secureSymlinkChown` and `mountSecretFs` functions. That probably shouldn't be too much work, but I can't claim to understand all the security details of those functions. Does anyone else who's more familiar with the code have access to a Mac?

**Comment by pogobanane** [comment]
i started a darwin home-manager implementation based on this PR here: https://github.com/pogobanane/sops-nix/tree/feat/home-manager-darwin 
- which compiles
- whose git history needs to be cleaned up
- which is completely untested
- which needs error handling
- whose workarounds/TODOs need work

Also there is https://github.com/4825764518/sops-nix/commits/darwin which implements a darwinModule (nixosModule for MacOs, _not_ home-manager module).

**Comment by mtoohey31** [comment]
Sorry for the slow reply, your fork works for me! I made two small changes [here](https://github.com/mtoohey31/sops-nix/tree/feat/home-manager-darwin), and with those changes I was able to set up a secret for my github cli credentials on [this branch](https://github.com/mtoohey31/infra/tree/feat/home-manager-sops) of my personal configuration. The secret is being set up properly on reboot too.

Aside from that, I haven't done any extensive testing. I'm not familiar with all of `sops-nix`'s features, so feel free to add on if I've missed anything, but here's some of the cases I figure I should still check:

- [ ] Different formats
- [ ] Different modes
- [ ] Non-default `sopsFile` values

**Comment by Mic92** [comment]
Does nix-darwin supports some sort of unitests? I am not a regular macos user and I fear I might break it accidentally.

**Comment by mtoohey31** [comment]
> Does nix-darwin supports some sort of unitests? I am not a regular macos user and I fear I might break it accidentally.

Tests would probably be a good idea. Both nix-darwin and home-manager have tests included in their main repositories. nix-darwin's are kind of ad-hoc and not really extensible as far as I can tell, but home-manager uses [nmt](https://gitlab.com/rycee/nmt); it might be helpful here.

**Comment by mtoohey31** [comment]
I've finished some more testing; it looks like all the things I had listed on the checklist above are working properly. I would recommend setting an output path for the launchd agent's logs though, as I have in d6eb1b8e4d95285a8c1843920dbb7554b22f949a.

**Comment by anilanar** [comment]
I'm waiting for it to be exposed in flakes to test it.

**Comment by dasJ** [comment]
Feel free to open a PR against this branch if you need this feature.

**Comment by anilanar** [comment]
@dasJ Will do so once I manage to make it work.

Right now, systemd service fails to start. One thing I noticed is that there's a missing space somewhere:

```
> cat /nix/store/jyk79akz7j478pr2f4kfg9pid0b88vs5-sops-nix-user
#!/nix/store/iffl6dlplhv22i2xy7n1w51a5r631kmi-bash-5.1-p16/bin/bash
export SOPS_GPG_EXEC=/nix/store/2ypi8rh1nm1v2w2pafva4rk1vigsjjvj-gnupg-2.3.4/bin/gpg/nix/store/7zi58cdmn8csqi59sp91f1f3bwi9lwcb-sops-install-secrets-0.0.1/bin/sops-install-secrets -ignore-passwd '/nix/store/1nf3f2pssq0nj5p4frcx0yhrplbpqxjq-manifest.json'
```

There's no space between `gpg` and `sops-install-secrets` paths.

**Comment by otavio** [comment]
Is anything still pending? I look forward to this feature.

**Comment by Mic92** [comment]
Related https://github.com/jordanisaacs/homeage

**Comment by anilanar** [comment]
I've been using this branch with gpg for a while now, with some fixes applied at https://github.com/Mic92/sops-nix/compare/master...anilanar:sops-nix:feat/home-manager-flake.

Maybe merge and wait for user feedback? I don't see any way to make progress but wait for people to test this if it works.

**Comment by Mic92** [comment]
@anilanar sure, just open a new PR and I will review it.

**Comment by anilanar** [comment]
Let's merge this? One minor note though: this PR doesn't fix #163 unlike what PR description says, if I'm not mistaken.

**Comment by G3zz** [comment]
Are there any blockers on this now?

**Comment by pogobanane** [comment]
i merged the home-manager-darwin version with this branch in #261. @mtoohey31 could you do another quick test if it still works on darwin?

**Comment by Mic92** [comment]
Closing this one. since the @pogobanane one is now more advanced.

**Comment by mtoohey31** [comment]
> @mtoohey31 could you do another quick test if it still works on darwin?

Sorry that I never replied to this. I see here: https://github.com/Mic92/sops-nix/pull/261#issuecomment-1412683822 that you were able to test things on darwin, so I take it there's no need for me to do any testing at this point?

**Pull request closed without merging**
